firefox_ui_harness/testcases/update.py

```diff
@@ -0,0 +1,234 @@
+            about_window.download()
+            about_window.wait_for_update_applied()
+
+        finally:
+            self.updates[self.current_update_index]['patch'] = about_window.patch_info
+
+        if force_fallback:
+            # Set the downloaded update into failed state
+            self.software_update.force_fallback()
+
+        # Restart Firefox to apply the downloaded update
+        self.restart()
+
+    def download_and_apply_forced_update(self):
+        # The update wizard dialog opens automatically after the restart
+        dialog = self.windows.switch_to(lambda win: type(win) is UpdateWizardDialog)
+
+        # In case of a broken complete update the about window has to be used
+        if self.updates[self.current_update_index]['patch']['is_complete']:
+            try:
+                self.assertEqual(dialog.wizard.selected_panel,
+                                 dialog.wizard.error)
+                dialog.close()
+
+                # Open the about window and check for updates
+                about_window = self.browser.open_about_window()
+                update_available = about_window.check_for_updates()
+                self.assertTrue(update_available)
+
+                # Download update and wait until it has been applied
+                about_window.download()
+                about_window.wait_for_update_applied()
+
+            finally:
+                self.updates[self.current_update_index]['patch'] = about_window.patch_info
+
+        else:
+            try:
+                self.assertEqual(dialog.wizard.selected_panel,
+                                 dialog.wizard.error_patching)
+
+                # Start downloading the fallback update
+                dialog.download()
+                dialog.close()
+
+            finally:
+                self.updates[self.current_update_index]['patch'] = dialog.patch_info
+
+        # Restart Firefox to apply the update
+        self.restart()
+
+    def restore_config_files(self):
+        # Reset channel-prefs.js file if modified
+        try:
+            if self.default_update_channel:
+                path = self.default_update_channel['path']
+                self.logger.info('Restoring channel defaults for: {}'.format(path))
+                with open(path, 'w') as f:
+                    f.write(self.default_update_channel['content'])
+        except IOError:
+            self.logger.error('Failed to reset the default update channel.',
+                              exc_info=True)
+
+        # Reset update-settings.ini file if modified
+        try:
+            if self.default_mar_channels:
+                path = self.default_mar_channels['path']
+                self.logger.info('Restoring mar channel defaults for: {}'.format(path))
+                with open(path, 'w') as f:
+                    f.write(self.default_mar_channels['content'])
+        except IOError:
+            self.logger.exception('Failed to reset the default mar channels.',
```

Comment:
Damn. This somehow got back. I clearly fixed that earlier. Thanks for spotting it.